[state_dict][5/N] Add submodules save and load support

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111111
* #111120
* __->__ #111110
* #111109
* #111275
* #111107
* #111106

It is not easy for user to do submodules save and load (e.g., fine tuning) because FSDP requires to get the root module. This PR enables the support of submodule save and load.

Differential Revision: [D50209727](https://our.internmc.facebook.com/intern/diff/D50209727/)

fegin committed:
[state_dict][5/N] Add submodules save and load support

It is not easy for user to do submodules save and load (e.g., fine tuning) because FSDP requires to get the root module. This PR enables the support of submodule save and load.

Differential Revision: [D50209727](https://our.internmc.facebook.com/intern/diff/D50209727/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][5/N] Add submodules save and load support"


It is not easy for user to do submodules save and load (e.g., fine tuning) because FSDP requires to get the root module. This PR enables the support of submodule save and load.

Differential Revision: [D50209727](https://our.internmc.facebook.com/intern/diff/D50209727/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][5/N] Add submodules save and load support"


It is not easy for user to do submodules save and load (e.g., fine tuning) because FSDP requires to get the root module. This PR enables the support of submodule save and load.

Differential Revision: [D50209727](https://our.internmc.facebook.com/intern/diff/D50209727/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][5/N] Add submodules save and load support"


It is not easy for user to do submodules save and load (e.g., fine tuning) because FSDP requires to get the root module. This PR enables the support of submodule save and load.

Differential Revision: [D50209727](https://our.internmc.facebook.com/intern/diff/D50209727/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][5/N] Add submodules save and load support"


It is not easy for user to do submodules save and load (e.g., fine tuning) because FSDP requires to get the root module. This PR enables the support of submodule save and load.

Differential Revision: [D50209727](https://our.internmc.facebook.com/intern/diff/D50209727/)

[ghstack-poisoned]

wz337 reviewed (approved):
Accepted in phabricator.

fegin committed:
Update on "[state_dict][5/N] Add submodules save and load support"


It is not easy for user to do submodules save and load (e.g., fine tuning) because FSDP requires to get the root module. This PR enables the support of submodule save and load.

Differential Revision: [D50209727](https://our.internmc.facebook.com/intern/diff/D50209727/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][5/N] Add submodules save and load support"


It is not easy for user to do submodules save and load (e.g., fine tuning) because FSDP requires to get the root module. This PR enables the support of submodule save and load.

Differential Revision: [D50209727](https://our.internmc.facebook.com/intern/diff/D50209727/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][5/N] Add submodules save and load support"


It is not easy for user to do submodules save and load (e.g., fine tuning) because FSDP requires to get the root module. This PR enables the support of submodule save and load.

Differential Revision: [D50209727](https://our.internmc.facebook.com/intern/diff/D50209727/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][5/N] Add submodules save and load support"


It is not easy for user to do submodules save and load (e.g., fine tuning) because FSDP requires to get the root module. This PR enables the support of submodule save and load.

Differential Revision: [D50209727](https://our.internmc.facebook.com/intern/diff/D50209727/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][5/N] Add submodules save and load support"


It is not easy for user to do submodules save and load (e.g., fine tuning) because FSDP requires to get the root module. This PR enables the support of submodule save and load.

Differential Revision: [D50209727](https://our.internmc.facebook.com/intern/diff/D50209727/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][5/N] Add submodules save and load support"


It is not easy for user to do submodules save and load (e.g., fine tuning) because FSDP requires to get the root module. This PR enables the support of submodule save and load.

Differential Revision: [D50209727](https://our.internmc.facebook.com/intern/diff/D50209727/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][5/N] Add submodules save and load support"


It is not easy for user to do submodules save and load (e.g., fine tuning) because FSDP requires to get the root module. This PR enables the support of submodule save and load.

Differential Revision: [D50209727](https://our.internmc.facebook.com/intern/diff/D50209727/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][5/N] Add submodules save and load support"


It is not easy for user to do submodules save and load (e.g., fine tuning) because FSDP requires to get the root module. This PR enables the support of submodule save and load.

Differential Revision: [D50209727](https://our.internmc.facebook.com/intern/diff/D50209727/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge
